src/lib.rs

```diff
@@ -6,73 +6,66 @@
-//!            detect_encoding_name(&[0x46, 0x72, 0x61, 0x6e, 0xe7, 0x6f, 0x69, 0x73, 0xe9]));
+//! assert_eq!("ISO-8859-1".to_string(),
+//!            detect_encoding_name(&[0x46, 0x72, 0x61, 0x6e, 0xe7, 0x6f,
+//!                0x69, 0x73, 0xe9]).unwrap());
 //! ```
 //!
 //! For more information, see [this project on
 //! GitHub](https://github.com/emk/rust-uchardet).
 
+// Increase the compiler's recursion limit for the `error_chain` crate.
+#![recursion_limit = "1024"]
 #![deny(missing_docs)]
 
+#[macro_use]
+extern crate error_chain;
 extern crate libc;
 extern crate uchardet_sys as ffi;
 
 use libc::size_t;
-use std::error::Error;
-use std::fmt;
-use std::result::Result;
 use std::ffi::CStr;
 use std::str::from_utf8;
 
-/// An error occurred while trying to detect the character encoding.
-#[derive(Debug, PartialEq)]
-pub struct EncodingDetectorError {
-    message: String
-}
-
-impl Error for EncodingDetectorError {
-    fn description(&self) -> &str { "encoding detector error" }
-    fn cause(&self) -> Option<&Error> { None }
-}
+use errors::*;
 
-impl fmt::Display for EncodingDetectorError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", &self.message)
+mod errors {
+    error_chain! {
+        errors {
+            UnrecognizableCharset {
+                description("unrecognizable charset")
+                display("uchardet was unable to recognize a charset")
+            }
+            DataHandlingError {
+                description("data handling error")
+                display("uchardet failed to handle the supplied data (out of memory?)")
+            }
+        }
     }
 }
 
-/// Either a return value, or an encoding detection error.
-pub type EncodingDetectorResult<T> = Result<T, EncodingDetectorError>;
-
 /// Detects the encoding of text using the uchardet library.
 ///
 /// EXPERIMENTAL: This may be replaced by a better API soon.
 struct EncodingDetector {
     ptr: ffi::uchardet_t
 }
 
-/// Return the name of the charset used in `data`, or `None` if the
-/// charset if the encoding can't be detected. This is
-/// the value returned by the underlying `uchardet` library, with
-/// the empty string mapped to `None`.
+/// Return the name of the charset used in `data` or an error if uchardet
+/// was unable to detect a charset.
 ///
 /// ```
 /// use uchardet::detect_encoding_name;
 ///
-/// assert_eq!(Ok("ASCII".to_string()),
-///            detect_encoding_name("ascii".as_bytes()));
-/// assert_eq!(Ok("UTF-8".to_string()),
-///            detect_encoding_name("©français".as_bytes()));
-/// assert_eq!(Ok("ISO-8859-1".to_string()),
-///            detect_encoding_name(&[0x46, 0x72, 0x61, 0x6e, 0xe7, 0x6f, 0x69, 0x73, 0xe9]));
-
-
-
+/// assert_eq!("ASCII".to_string(),
+///            detect_encoding_name("ascii".as_bytes()).unwrap());
+/// assert_eq!("UTF-8".to_string(),
+///            detect_encoding_name("©français".as_bytes()).unwrap());
+/// assert_eq!("ISO-8859-1".to_string(),
```

Comment:
These `to_string` calls in these assertions are no longer necessary now that the surrounding `Ok` has been removed from the strings, because you can compare `&str` and `String`, but not `Ok(&str)` and `Ok(String)`, if I recall correctly.